common/quotas/policy.go

```diff
@@ -43,3 +49,57 @@ func (s *simpleRateLimitPolicy) Allow() bool {
 func (s *simpleRateLimitPolicy) Wait(d time.Duration) bool {
 	return s.tb.Consume(1, d)
 }
+
+// DomainRateLimitPolicy indicates a domain specific rate limit policy
+type DomainRateLimitPolicy struct {
+	sync.RWMutex
+	rps            dynamicconfig.IntPropertyFn
+	domainLimiters map[string]*rate.Limiter
+	globalLimiter  atomic.Value
+}
+
+// NewDomainRateLimiter returns a new domain quota rate limiter
+func NewDomainRateLimiter(rps dynamicconfig.IntPropertyFn) *DomainRateLimitPolicy {
+	return newDomainRateLimiter(rps())
+}
+
+func newDomainRateLimiter(rps int) *DomainRateLimitPolicy {
+	rl := &DomainRateLimitPolicy{
+		domainLimiters: map[string]*rate.Limiter{},
+	}
+	rl.globalLimiter.Store(rate.NewLimiter(rate.Limit(rps), defaultRps))
+	return rl
+}
+
+func (d *DomainRateLimitPolicy) Wait(dur time.Duration) bool {
+	return false
+}
+
+func (d *DomainRateLimitPolicy) Allow(domain string) bool {
+	// check if we have a per-domain limiter - if not create a default one for
+	// the domain.
+	d.RLock()
+	limiter, ok := d.domainLimiters[domain]
+	d.RUnlock()
+	if !ok {
+		d.Lock()
+		limiter = rate.NewLimiter(rate.Limit(defaultRps), defaultRps)
```

Comment:
1. this line can be out of the lock
2. before putting the limiter into the map, do a double check?